File: hw10/systemf.scala
import collection.mutable.HashMap;

type Var = String;
type TyVar = String;

enum Term {
  case Variable(v:Var)
  case Lambda(v:Var,annot:Ty,body:Term)
  case App(fun:Term,arg:Term)
  case TyLambda(ty: TyVar, term: Term)
  case TyApp(term: Term, ty: Ty)  
}

enum Ty {
  case VarTy(v:TyVar)
  case FunTy(from: Ty, to: Ty)
  case ForallTy(v:TyVar, ty: Ty)
}

case class Env(vars:HashMap[Var,Ty],tyvars:List[TyVar])
def emptyEnv() : Env = {
  Env(HashMap(),List())
}

def check(t: Term, env: Env) : Option[Ty] = {
  None // TODO
}

@main
def main() = {
  import Term.*
  import Ty.*
  val id = TyLambda("X",Lambda("x", VarTy("X"), Variable("x")));
  val id_ty = check(id, emptyEnv())
  assert(id_ty==ForallTy("X",FunTy(VarTy("X"),VarTy("X"))))

  val double = TyLambda("X", Lambda("f", FunTy(VarTy("X"),VarTy("X")), Lambda("a",VarTy("X"),App(Variable("f"), App(Variable("f"),Variable("a"))))))
  val double_ty = check(double, emptyEnv());

  assert(double_ty==ForallTy("X", FunTy(FunTy(VarTy("X"),VarTy("X")),FunTy(VarTy("X"),VarTy("X")))));
}

File: hw10/systemf.rs
use std::collections::HashMap;

pub type Var = String;
pub type TyVar = String;

pub enum Term {
    Var(Var),
    Lambda {
        var: Var,
        annot: Type,
        body: Box<Term>,
    },
    App {
        fun: Box<Term>,
        arg: Box<Term>,
    },
    TyLambda {
        ty: TyVar,
        term: Box<Term>,
    },
    TyApp {
        term: Box<Term>,
        ty: Type,
    },
}

#[derive(Debug, PartialEq, Eq)]
pub enum Type {
    Var(TyVar),
    Fun { from: Box<Type>, to: Box<Type> },
    Forall { var: TyVar, ty: Box<Type> },
}

#[derive(Default)]
pub struct Env {
    vars: HashMap<Var, Type>,
    ty_vars: Vec<TyVar>,
}

pub fn check(t: Term, env: &mut Env) -> Option<Type> {
    todo!()
}

fn main() {
    let id = Term::TyLambda {
        ty: "X".to_owned(),
        term: Box::new(Term::Lambda {
            var: "x".to_owned(),
            annot: Type::Var("X".to_owned()),
            body: Box::new(Term::Var("x".to_owned())),
        }),
    };
    let id_ty = check(id, &mut Default::default()).unwrap();
    assert_eq!(
        id_ty,
        Type::Forall {
            var: "X".to_owned(),
            ty: Box::new(Type::Fun {
                from: Box::new(Type::Var("X".to_owned())),
                to: Box::new(Type::Var("X".to_owned()))
            })
        }
    );

    let double = Term::TyLambda {
        ty: "X".to_owned(),
        term: Box::new(Term::Lambda {
            var: "f".to_owned(),
            annot: Type::Fun {
                from: Box::new(Type::Var("X".to_owned())),
                to: Box::new(Type::Var("X".to_owned())),
            },
            body: Box::new(Term::Lambda {
                var: "a".to_owned(),
                annot: Type::Var("X".to_owned()),
                body: Box::new(Term::App {
                    fun: Box::new(Term::Var("f".to_owned())),
                    arg: Box::new(Term::App {
                        fun: Box::new(Term::Var("f".to_owned())),
                        arg: Box::new(Term::Var("a".to_owned())),
                    }),
                }),
            }),
        }),
    };
    let double_ty = check(double, &mut Default::default()).unwrap();
    assert_eq!(
        double_ty,
        Type::Forall {
            var: "X".to_owned(),
            ty: Box::new(Type::Fun {
                from: Box::new(Type::Fun {
                    from: Box::new(Type::Var("X".to_owned())),
                    to: Box::new(Type::Var("X".to_owned())),
                }),
                to: Box::new(Type::Fun {
                    from: Box::new(Type::Var("X".to_owned())),
                    to: Box::new(Type::Var("X".to_owned())),
                }),
            }),
        }
    );
}
